service/storage_proxy.cc

```diff
@@ -3646,6 +3646,11 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
     virtual future<foreign_ptr<lw_shared_ptr<query::result>>> execute(storage_proxy::clock_type::time_point timeout) {
         digest_resolver_ptr digest_resolver = ::make_shared<digest_read_resolver>(_schema, _cl, _block_for,
                 db::is_datacenter_local(_cl) ? db::count_local_endpoints(_targets): _targets.size(), timeout);
+        if (timeout <= storage_proxy::clock_type::now()) {
+            return make_exception_future<foreign_ptr<lw_shared_ptr<query::result>>>(
+                exceptions::overloaded_exception(format("Request timed out before being sent to replicas"))
```

Comment:
On a timeout `timeout_exception` should be issued.

Reply:
`format` was a leftover from a more sophisticated message, it's redundant now. As for the exception type - I don't agree, I think that clients would be able to apply more specific retry policies with `overloaded`. Timeout is very generic and doesn't inform the user what actually timed out - and it should usually happen after the request is already passed to replicas. If we timed out before having a chance of sending anything, that's a clear sign that the coordinator is overloaded and it would make sense to try on another one (which is how drivers usually react on `OverloadedException`. The most fitting error type would be `CoordinatorTimeout`, but CQL doesn't have that one specified.

Reply:
For instance, the Python driver prints such a description for read timeout:
```
cassandra.ReadTimeout: Error from server: code=1200 [Coordinator node timed out waiting for replica nodes' responses]
```
It's not strictly true, and should be fixed on the driver side, but the fact that everybody accepted this description also shows how people treat read timeouts. The fact that this timeout error has slots for "how many replicas were needed" and "how many replicas responded" also doesn't match well if it's the coordinator node that hasn't managed to send even a single request.

Reply:
You do have the information for "how many replicas were needed" and "how many replicas responded" here. The later is zero of course since no replicas were ever contacted in time. I can see how you may want to consider a case when timeout happens even before the request is started to be processed as an overload, but this may happen even without overload if timeout is just too small.